[crypto] allow platform to define its own CSPRNG

LuDuda opened the pull request:
This PR provides the following changes:

- Introduce a new otPlatCryptoRandom API
- Move usage of `mbedtls_entropy_*` and `mbedtls_ctr_drbg_*` out of ARM PSA enabled platforms
- Stop exposing internal mbedtls contextes for entropy and CTR DRBG in OpenThread API

Background: With TF-M enabled, access to secure random number generation should be done only via `psa_random_get` function. Non-secure environment should not use crypto HW directly (which might be the case by using legacy mbedtls entropy and ctr_drbg functions.

Signed-off-by: Lukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda committed:
[crypto] allow platform to define its own CSPRNG

This commit provides the following changes:

 * Introduce a new otPlatCryptoRandom API
 * Move usage of mbedtls_entropy_* and mbedtls_ctr_drbg_* out of
   ARM PSA enabled platforms
 * Stop exposing internal mbedtls contextes for entropy and CTR DRBG
   in OpenThread API

Signed-off-by: Lukasz Duda <lukasz.duda@nordicsemi.no>

LuDuda commented:
FYI @Damian-Nordic

Damian-Nordic commented on src/core/crypto/crypto_platform.cpp:
```diff
@@ -419,4 +435,53 @@ OT_TOOL_WEAK otError otPlatCryptoSha256Finish(otCryptoContext *aContext, uint8_t
+
+OT_TOOL_WEAK otError otPlatCryptoRandomDeinit(void)
+{
+    mbedtls_ctr_drbg_free(&sCtrDrbgContext);
```
What about `sEntropyContext`?

Damian-Nordic commented on src/core/crypto/crypto_platform.cpp:
```diff
@@ -51,9 +54,22 @@
+#endif
+
+#if !OPENTHREAD_RADIO
+static mbedtls_ctr_drbg_context sCtrDrbgContext;
```
Wouldn't it be more future-proof if the context was passed via `otCryptoContext` like in other crypto functions? For example, what if a single application runs multiple OT instances?

Damian-Nordic replied:
Or you did so, because `Random::Crypto::MbedTlsContextGet()` is global anyway?

LuDuda replied:
Yes, currently there are no multi-instances of RNGs algorithms in OT. Let me maybe push this change and explicitly ask OT community whether they envision a use-case for the future, and if so, i will add `otCryptoContext` param to `otPlatCryptoRandom*` functions.

zhanglongxia committed:
[posix] add `break` to case statement (#7349)

simonlingoogle committed:
[github-actions] add BR `--radio-version` test (#7345)

JaneFromSilabs committed:
[thci] reset and multicast changes (#7340)

- THCI changes for bbr reset procedure to make a better effort to
  reset and to also make sure to restart the otbr-agent prior to
  attempting a factoryreset to cover the case where the agent cannot
  currently communicate to the rcp.

- THCI change to move from sudo service restart otbr-agent to sudo
  systemctl restart otbr-agent, which is more comprehensive.

- THCI changes for multicast to use ipmaddr in both soc and bbr cases.

edmont committed:
[bbr] fix `bbr skipseqnuminc` command response (#7354)

`bbr skipseqnuminc` should complete without error.

abtink committed:
[mac] smaller style changes (#7341)

This commit contain smaller style changes in `Mac` and `SubMac`

abtink committed:
[cli] inline simple sub-command process methods (#7347)

abtink committed:
[cli] add helper method to output a `otMacFilterEntry` (#7348)

This commit also adds `MacFilterAddressModeToString()`.

gjc13 committed:
[ip6] always enable ICMP6 filtering when forwarding to Thread (#7343)

This will prevent unexpected multicast packets on the Thread network.

abtink committed:
[address-resolver] add `LookUp()` method to check cache entries (#7350)

This commit adds `AddressResolver::LookUp()` method which resolves
an EID to an RLOC17 by checking if an existing entry already exists
in the address cache table.

edmont committed:
[bbr] implement new Sequence Number increase rules (#7342)

"BBR Sequence Number Updating" section in Thread Specification
requires special wraps when increasing BBR Dataset Sequence Number.

abtink committed:
[commissioner] add `Dataset` and `ResignMode` (#7352)

This commit adds `Commissioner::Dataset` class which mirrors the
public `otCommissioningDataset`. It also adds `ResignMode` enum
which is used as input to `Stop()` method to indicate whether or
not to resign from being the commissioner.

LuDuda committed:
Address review comment

xuyirio committed:
[nat64] withdraw greater NAT64 prefix in network data (#7337)

This commit enables BR to withdraw its local NAT64 prefix if a smaller
one is found in network data.

It might happen when two BRs join the Thread network at merely the
same time and thus both advertise its local NAT64 prefix. The BR with
numerically greater prefix will withdraw its NAT64 prefix.

wohdo committed:
[coap] add 'otCoapMessageSetCode' (#7339) (#7351)

Provide a setter for the message code to change it after
initialization of the message.

This complements the existing getter function `otCoapMessageGetCode`.

LuDuda committed:
Change return types

abtink committed:
[logging] remove `otLogResult` and move to module (#7360)

This commit removes the `otLogResult{Bbr/Platform}()` macros from
`logging.hpp` and instead have each module implement `LogError()`
method. The main goal behind this change to simplify the logging
module and reduce number of macros/definitions in it (to prepare for
new logging model to be introduced).

bukepo committed:
[cli] do not update aError which is not read (#7365)

This commit avoid setting aError which is not read.
